lib/galaxy/tools/execute.py

```diff
@@ -211,7 +211,32 @@ def output_name(self, trans, history, params, output):
         return output_collection_name
 
     def sliced_input_collection_structure(self, input_name):
-        input_collection = self.example_params[input_name]
+        def find_collection(input_dict, input_name):
+            for key, value in input_dict.items():
+                is_dict = isinstance(value, dict)
+                if key == input_name:
+                    return value
+                if "|" in input_name and is_dict:
+                    prefix, rest_input_name = input_name.split("|", 1)
+                    if key == prefix:
+                        return find_collection(value, rest_input_name)
+
+                if is_dict and self.tool.profile < 18.09:
+                    # Looking for "input1" instead of "cond|input1" for instance.
+                    # See discussion on https://github.com/galaxyproject/galaxy/issues/6157.
+                    unqualified_match = find_collection(value, input_name)
+                    if unqualified_match:
+                        return unqualified_match
```

Comment:
As it is now, we may continue recursing after removing some prefixes from `input_name` if the tool has a profile < 18.09. What about this?
```python
    def sliced_input_collection_structure(self, input_name):
        unqualified_recurse = self.tool.profile < 18.09 and "|" not in input_name
        def find_collection(input_dict, input_name):
            for key, value in input_dict.items():
                if key == input_name:
                    return value
                if isinstance(value, dict):
                    if "|" in input_name:
                        prefix, rest_input_name = input_name.split("|", 1)
                        if key == prefix:
                            return find_collection(value, rest_input_name)
                    elif unqualified_recurse:
                        # Looking for "input1" instead of "cond|input1" for instance.
                        # See discussion on https://github.com/galaxyproject/galaxy/issues/6157.
                        unqualified_match = find_collection(value, input_name)
                        if unqualified_match:
                            return unqualified_match
```